common/minimodel/__init__.py

```diff
@@ -1008,3 +1013,52 @@ def total_used(self) -> timedelta:
 
     def observations(self) -> List[Observation]:
         return self.root_group.observations()
+
+
+class Plans:
+    """
+    A collection of Nightly Plan from a corresponding Site
+    """
+    def __init__(self, night_events):
+        self.durations = [len(night_events.times[idx]) for idx, _ in enumerate(night_events.time_grid)]
+        self.start_times = [night_events.local_times[idx][0] for idx, _ in enumerate(night_events.time_grid)]
+        self.end_times = [night_events.local_times[idx][-1] for idx, _ in enumerate(night_events.time_grid)]
+        self.time_slot_length = night_events.time_slot_length
+
+    def __iter__(self):
+        self.plans = []
+        return self
+
+    def __next__(self):
+        if len(self.plans) <= len(self.durations):
+            self.plans.append(Plan(next(iter(self.start_times)),
+                                   next(iter(self.end_times)),
+                                   slot_length=self.time_slot_length,
+                                   night_duration=next(iter(self.durations))))
+            return next(iter(self.plans))
+        raise StopIteration
+
+
+class Plan:
+    """
+    A 'plan' is a collection of nighly plans
+    """
+    def __init__(self, start, end, slot_length=1, night_duration=10):
+        self.start = start
+        self.end = end
+        self._time_slot_length = slot_length
+        self._time_slots_left = night_duration
+        self._visits = []
+    
+    def _time2slots(self, time) -> int:
+        return ceil((time.total_seconds() / 60) / self._time_slot_length.value)
+
+    def add_group(self, group):
+        self._visits.append(group)
+        self._time_slots_left -= self._time2slots(group[0].total_used())
```

Comment:
`group` might be a bit confusing  name convention as the parameter is not a Group object but a pair `[Group, GroupInfo]`. So that access the Group object in fact

Reply:
Got it. I see it now with the type annotation and it makes sense to me now.
Let's talk about the concept of nightly plans when we meet.

Reply:
This also limits us in that we will be adding visits sequentially to the plan, which may cause problems. I don't think that is the behavior we want, since GreedyMax is not always going to schedule from start to end, and any other optimizer (if we end up writing another non-dummy one) will likely not schedule from start to end either.

This is why I was thinking of the simpler implementation of the concept of plans as per above.